### PR TITLE
Eclair rpc v2

### PR DESCRIPTION
**Pull request by Christewart**
This builds off of #188 

This PR aims to implement a RPC for [eclair](https://github.com/acinq/eclair) -- a lightning network implementation. 

We are aiming to have a full RPC client with a comprehensive set of unit tests that open create their own eclair instances, run unit tests against those eclair instances, and then destroy them.

Currently we have a unit test implemented for `open`

Here is a full list of Eclair's RPCs

https://github.com/acinq/eclair#json-rpc-api

### CODE REVIEW

File: core-test/src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
     </appender>
 
 
-    <root level="INFO">
+    <root level="OFF">
```
Comment (nkohen) [inline core-test/src/test/resources/logback-test.xml]: Not even ERROR for CI purposes?

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  override def beforeAll(): Unit = {
+    logger.info("Temp eclair directory created")
+    logger.info("Temp eclair directory created")
+
+    logger.debug(client.getDaemon.authCredentials.datadir.toString)
+
+    client.start()
+    otherClient.start()
+
+    Thread.sleep(20000)
```
Comment (nkohen) [inline eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala]: We should probably use the RpcUtil stuff

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    val infoF = otherClient.getInfo
+
+    val connected: Future[String] = infoF.flatMap { info =>
+      client.connect(info.nodeId, "localhost", info.port)
+    }
+
+    Await.result(connected, 15.seconds)
```
Comment (nkohen) [inline eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala]: RpcUtil here as well

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  private def hasChannel(client: EclairRpcClient, chanId: String): Future[Assertion] = {
+    val recognizedOpenChannel: Future[Assertion] = {
+      val chanResultF: Future[ChannelResult] = client.channel(chanId)
+      chanResultF.map(c => assert(c.channelId == chanId))
+    }
+
+    recognizedOpenChannel
+  }
+
+  override def afterAll(): Unit = {
```
Comment (nkohen) [inline eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala:94]: Should probably terminate the ActorSystem here as well

File: .travis.yml [hunk 5]
```diff
@@ -36,6 +36,9 @@ install:
   - ./configure --enable-jni --enable-experimental --enable-module-ecdh
   - sudo make install
   - cd ../
+  - wget https://github.com/ACINQ/eclair/releases/download/v0.2-beta5/eclair-node-0.2-beta5-8aa51f4.jar
+  - export ECLAIR_PATH=$(pwd)
```
Comment (nkohen) [inline .travis.yml:40]: Document this?

**Review by nkohen — CHANGES_REQUESTED** [review]

File: eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/client/EclairRpcClient.scala [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+  def start(): Unit = {
+    val p = Process("java -jar -Declair.datadir=" + instance.authCredentials.datadir + s" ${pathToEclairJar} &")
+    val result = p.run()
+    logger.info(s"Starting eclair with datadir ${instance.authCredentials.datadir}")
+
+    process = Some(result)
+    ()
+  }
+
+  def isStarted(): Boolean = {
```
Comment (nkohen) [inline eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/client/EclairRpcClient.scala]: We could probably do a check for `process == None` before stalling the program that calls this for 1.second

File: eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/json/EclairModels.scala [hunk 7]
```diff
@@ -0,0 +1,188 @@
+package org.bitcoins.eclair.rpc.json
+
+import play.api.libs.json.{ JsArray, JsObject }
+
+sealed abstract class EclairModels
```
Comment (nkohen) [inline eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/json/EclairModels.scala:5]: Is there a reason that this isn't extended by anything in here? Also is there a difference between `abstract class` and `trait` in this instance?

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    logger.info("Temp eclair directory created")
+    logger.info("Temp eclair directory created")
+
+    logger.debug(client.getDaemon.authCredentials.datadir.toString)
+
+    client.start()
+    otherClient.start()
+
+    RpcUtil.awaitCondition(
+      condition = () => client.isStarted(),
```
Comment (nkohen) [inline eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala]: are these`() =>` necessary? I think it's a bit more reader-friendly to drop them

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    recognizedOpenChannel
+  }
+
+  override def afterAll(): Unit = {
+    val s1 = client.stop()
+    val s2 = otherClient.stop()
+    val s3 = bitcoindRpc.stop()
+    system.terminate()
```
Comment (nkohen) [inline eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala:98]: This returns a Future we should probably block on

File: rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala [hunk 10]
```diff
@@ -68,25 +78,48 @@ trait RpcUtil extends BitcoinSLogger {
    * @param system An ActorSystem to schedule calls to condition
    */
   def awaitCondition(
-    condition: => Boolean,
+    condition: () => Boolean,
+    duration: FiniteDuration = 100.milliseconds,
+    maxTries: Int = 50,
+    overallTimeout: FiniteDuration = 1.hour)(implicit system: ActorSystem): Unit = {
+    implicit val ec = system.dispatcher
+    def h: Boolean = condition()
```
Comment (nkohen) [inline rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala]: why is this needed? I feel like `condition` is a more descriptive name and should just be used where `h` is

File: rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala [hunk 11]
```diff
@@ -17,42 +17,52 @@ trait RpcUtil extends BitcoinSLogger {
...
-   * @param condition The condition being waited on
+   * @param conditionF The condition being waited on
    * @param duration The interval between calls to check condition
    * @param maxTries If condition is tried this many times, the Future fails
    * @param system An ActorSystem to schedule calls to condition
    * @return A Future[Unit] that succeeds if condition becomes true and fails otherwise
    */
-  def retryUntilSatisfied(
-    condition: => Boolean,
+  def retryUntilSatisfiedF(
```
Comment (nkohen) [inline rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala:34]: I think it wouldn't hurt to also have a `retryUntilSatisfied` that takes in a `=> Boolean` with no `Future` which could literally be

```
def retryUntilSatisfiedF(
  condition: => Boolean,
  duration: FiniteDuration = 100.milliseconds,
  maxTries: Int = 50)(implicit system: ActorSystem): Future[Unit] = {
    retryUntilSatisfiedF(Future.successful(conditionF), duration, maxTries)
}

**Review by nkohen — CHANGES_REQUESTED** [review]

File: eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala (hunk 8 above)
Comment (Christewart) [reply eclair-rpc/src/test/scala/org/bitcoins/eclair/rpc/EclairRpcClientTest.scala]: I this case they are not necessary, but in the more general case they are. I.e. if `client.isStarted()` took params you would need to use an explicit anonymous function.

File: eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/json/EclairModels.scala (hunk 7 above)
Comment (Christewart) [reply eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/json/EclairModels.scala:5]: I default to `abstract class` over `trait` because it is "more native" to the JVM.

File: rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala (hunk 10 above)
Comment (Christewart) [reply rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala]: This is a bit of type hackery going on here. Basically we need to transform from a 

`() => Boolean`

to 

`() => Future[Boolean]`

the trick is figuring out how to do this WITHOUT caching the value -- but making it so that the `condition` is re-evaluated everytime you call `retryUntilSatisfied`.

Comment (nkohen) [reply rpc/src/main/scala/org/bitcoins/rpc/RpcUtil.scala]: Got it, a short comment and maybe the name `conditionCallByName` or `conditionDef` wouldn't hurt?
